soc/st/stm32/common/stm32_iocell.c

```diff
@@ -0,0 +1,218 @@
+#include <zephyr/dt-bindings/power/stm32h7rs_iocell.h>
+#elif defined(CONFIG_SOC_SERIES_STM32N6X)
+#include <zephyr/dt-bindings/power/stm32n6_iocell.h>
+#define STM32N6_IOCELL_FIXEDVALUE (0x00000287)
+#endif
+
+#define DT_DRV_COMPAT     st_stm32_iocell
+#define STM32_IOCELL_NODE DT_NODELABEL(iocell)
+
+/* Returns 1 if HSLV can be confgured */
+static inline int iocell_get_hslv_safeguard(uint16_t domain)
+{
+#if defined(CONFIG_SOC_SERIES_STM32H7RSX)
+	if (domain == STM32_DT_IOCELL_VDDIO) {
+		return (FLASH->OBW1SR & FLASH_OBW1SR_VDDIO_HSLV) == FLASH_OBW1SR_VDDIO_HSLV;
+	} else if (domain == STM32_DT_IOCELL_XSPI1) {
+		return (FLASH->OBW1SR & FLASH_OBW1SR_XSPI1_HSLV) == FLASH_OBW1SR_XSPI1_HSLV;
+	} else if (domain == STM32_DT_IOCELL_XSPI2) {
+		return (FLASH->OBW1SR & FLASH_OBW1SR_XSPI2_HSLV) == FLASH_OBW1SR_XSPI2_HSLV;
+	} else {
+		return -1;
+	}
+#elif defined(CONFIG_SOC_SERIES_STM32N6X)
+	static uint32_t hconf1 = 0xFFFFFFFF;
+	uint32_t shift = 16 - domain;
+
+	if (hconf1 == 0xFFFFFFFF) {
+		WRITE_REG(BSEC->OTPCR, 124);
+
+		while ((BSEC->OTPSR & BSEC_OTPSR_BUSY) != 0U) {
+		}
+
+		if (BSEC->OTPSR & (BSEC_OTPSR_DISTURBF | BSEC_OTPSR_DEDF | BSEC_OTPSR_AMEF)) {
+			/* Read error => force safe configuration */
+			hconf1 = 0x0FFFFFFF;
+		} else {
+			hconf1 = BSEC->FVRw[124];
+		}
+	}
+
+	if (domain == STM32_DT_IOCELL_VDD) {
+		shift = 17;
+	}
+	return (hconf1 & BIT(shift)) == BIT(shift);
+#endif
+}
+
+/* Enables HSLV*/
+static inline int iocell_enable_hslv_runtime(uint16_t domain)
+{
+#if defined(CONFIG_SOC_SERIES_STM32H7RSX)
+	if (domain == STM32_DT_IOCELL_VDDIO) {
+		LL_SBS_EnableIOSpeedOptim();
+	} else if (domain == STM32_DT_IOCELL_XSPI1) {
+		LL_SBS_EnableXSPI1SpeedOptim();
+	} else if (domain == STM32_DT_IOCELL_XSPI2) {
+		LL_SBS_EnableXSPI2SpeedOptim();
+	} else {
+		return -1;
+	}
+	return 0;
+#elif defined(CONFIG_SOC_SERIES_STM32N6X)
+	if (domain == STM32_DT_IOCELL_VDDIO2) {
+		LL_PWR_SetVddIO2VoltageRange(LL_PWR_VDDIO_VOLTAGE_RANGE_1V8);
+	} else if (domain == STM32_DT_IOCELL_VDDIO3) {
+		LL_PWR_SetVddIO3VoltageRange(LL_PWR_VDDIO_VOLTAGE_RANGE_1V8);
+	} else if (domain == STM32_DT_IOCELL_VDDIO4) {
+		LL_PWR_SetVddIO4VoltageRange(LL_PWR_VDDIO_VOLTAGE_RANGE_1V8);
+	} else if (domain == STM32_DT_IOCELL_VDDIO5) {
+		LL_PWR_SetVddIO5VoltageRange(LL_PWR_VDDIO_VOLTAGE_RANGE_1V8);
+	} else if (domain == STM32_DT_IOCELL_VDD) {
+		LL_PWR_SetVddIOVoltageRange(LL_PWR_VDDIO_VOLTAGE_RANGE_1V8);
+	}
+#endif
+}
+
+static inline void iocell_enable_compensation(uint16_t domain)
+{
+#if defined(CONFIG_SOC_SERIES_STM32H7RSX)
+	/* TODO: check MCU revision */
+
+	/* This implements workaround described in ES0596:
+	 * 2.2.15. I/O compensation could alter duty-cycle of high-frequency output signal
+	 */
+	uint32_t shift = domain;
+	uint32_t nmos, pmos;
+	/* TODO: skip if compensation cell already enabled ? */
+
+	/* Configure Cell selection for domain and enable */
+	MODIFY_REG(SBS->CCCSR, BIT((shift * 2) + 1), BIT(shift * 2));
```

Comment:
```suggestion
	stm32_reg_modify_bits(SBS->CCCSR, BIT((shift * 2) + 1), BIT(shift * 2));
```
See 004c613e25d957c35258132e2d60d1f20dc4b47f.